src/audio/copier/copier.c

```diff
@@ -780,6 +826,178 @@ static int copier_comp_trigger(struct comp_dev *dev, int cmd)
 	return ret;
 }
 
+static inline struct comp_buffer *get_endpoint_buffer(struct copier_data *cd)
+{
+	return cd->multi_endpoint_buffer ? cd->multi_endpoint_buffer :
+		cd->endpoint_buffer[IPC4_COPIER_GATEWAY_PIN];
+}
+
+static inline void copy_single_channel_c16(struct audio_stream __sparse_cache *dst,
+					   int dst_channel,
+					   const struct audio_stream __sparse_cache *src,
+					   int src_channel, int frame_count)
+{
+	int16_t *r_ptr = audio_stream_wrap(src, (int16_t *)src->r_ptr + src_channel);
+	int16_t *w_ptr = audio_stream_wrap(dst, (int16_t *)dst->w_ptr + dst_channel);
+	int i;
+
+	for (i = 0; i < frame_count; i++) {
+		*w_ptr = *r_ptr;
+		r_ptr = audio_stream_wrap(src, r_ptr + src->channels);
+		w_ptr = audio_stream_wrap(dst, w_ptr + dst->channels);
+	}
+}
+
+static inline void copy_single_channel_c32(struct audio_stream __sparse_cache *dst,
+					   int dst_channel,
+					   const struct audio_stream __sparse_cache *src,
+					   int src_channel, int frame_count)
+{
+	int32_t *r_ptr = audio_stream_wrap(src, (int32_t *)src->r_ptr + src_channel);
+	int32_t *w_ptr = audio_stream_wrap(dst, (int32_t *)dst->w_ptr + dst_channel);
+	int i;
+
+	for (i = 0; i < frame_count; i++) {
+		*w_ptr = *r_ptr;
+		r_ptr = audio_stream_wrap(src, r_ptr + src->channels);
+		w_ptr = audio_stream_wrap(dst, w_ptr + dst->channels);
+	}
+}
```

Comment:
don't we already have a library for such buffer copies?

Reply:
No, seems we do not have such functionality in the library. This is a coping of particular channel from one stream to another. Here it is used to multiplex or demultiplex streams according to channel map.